### DISCUSSION TITLE
createModel AssignAction is upsetting invoke

### DISCUSSION THREAD

**Discussion by tkofh** [statelyai/xstate#2022] [Q&A]
Hey, I believe this to be more of a typescript problem than an xstate question, but I'm hoping the error will be familiar to someone. I also realize that `createModel` is brand new, so I might be misusing anyways. 

In [this codesandbox example](https://codesandbox.io/s/xstate-createmodel-invoke-assign-ymj9m?file=/src/index.ts), I have a simple machine that invokes a promise, to then assign its results (or error message) to a model from `createModel`: 

```typescript
import { createMachine } from "xstate";
import { createModel } from "xstate/lib/model";

const fruitsAPI = (): Promise<string[]> =>
  Promise.resolve(["apples", "bananas"]);

interface FetchingContext {
  fruits?: string[];
  errorMessage?: string;
}

const fetchingModel = createModel(
  {
    fruits: undefined,
    errorMessage: undefined
  } as FetchingContext,
  {
    events: {
      "done.invoke.fetchFruits": (data: string[]) => ({ data }),
      "error.platform.fetchFruits": (data: string) => ({ data })
    }
  }
);

const fetchingMachine = createMachine<typeof fetchingModel>({
  initial: "fetching",
  context: fetchingModel.initialContext,
  states: {
    fetching: {
      invoke: {
        id: "fetchFruits",
        src: () => fruitsAPI(),
        onDone: {
          target: "success",
          actions: fetchingModel.assign<"done.invoke.fetchFruits">({
            fruits: (_, event) => event.data
          })
        },
        onError: {
          target: "error",
          actions: fetchingModel.assign<"error.platform.fetchFruits">({
            errorMessage: (_, event) => event.data
          })
        }
      }
    },
    success: {
      type: "final"
    },
    error: {
      type: "final"
    }
  }
});
```

In strict mode, typescript doesn't like the `invoke` key on the `fetching` state node for this reason: 
```
No overload matches this call.
  Overload 1 of 2, '(config: MachineConfig<FetchingContext, any, EventFromEventCreators<FinalEventCreators<{ "done.invoke.fetchFruits": (data: string[]) => { data: string[]; }; "error.platform.fetchFruits": (data: string) => { data: string; }; }>>>, options?: Partial<...> | undefined): StateMachine<...>', gave the following error.
    Type '{ id: string; src: () => Promise<string[]>; onDone: { target: string; actions: AssignAction<FetchingContext, { data: string[]; type: "done.invoke.fetchFruits"; }>; }; onError: { ...; }; }' is not assignable to type 'StateMachine<any, any, any, { value: any; context: any; }> | InvokeConfig<FetchingContext, EventFromEventCreators<FinalEventCreators<{ "done.invoke.fetchFruits": (data: string[]) => { data: string[]; }; "error.platform.fetchFruits": (data: string) => { ...; }; }>>> | (StateMachine<...> | InvokeConfig<...>)[] | undef...'.
      Types of property 'onDone' are incompatible.
        Type '{ target: string; actions: AssignAction<FetchingContext, { data: string[]; type: "done.invoke.fetchFruits"; }>; }' is not assignable to type 'string | TransitionConfig<FetchingContext, DoneInvokeEvent<any>> | TransitionConfig<FetchingContext, DoneInvokeEvent<any>>[] | undefined'.
          Types of property 'actions' are incompatible.
            Type 'AssignAction<FetchingContext, { data: string[]; type: "done.invoke.fetchFruits"; }>' is not assignable to type 'string | ActionObject<FetchingContext, DoneInvokeEvent<any>> | ActionFunction<FetchingContext, DoneInvokeEvent<any>> | Action<...>[] | undefined'.
              Type 'AssignAction<FetchingContext, { data: string[]; type: "done.invoke.fetchFruits"; }>' is not assignable to type 'string'.ts(2769)
```

If you remove `"strict": true` from the `tsconfig.json`, the error goes away.

I've tried a bunch of assertions in various places to no avail. Any guidance on how I can annotate the types better (or better utilize `createModel`) would be greatly appreciated.

Thanks!

**Comment by Andarist** [comment]
This happens because this is not assignable:
```ts
const test: ActionObject<typeof fetchingModel.initialContext, DoneInvokeEvent<any>> = fetchingModel.assign<"done.invoke.fetchFruits">({
  fruits: (_, event) => event.data
})
```

This most likely happens because of `strictFunctionTypes` setting as `ActionObject` has an optional `exec` property that is a function.

I'm not sure what we can do about it quickly though. I've already been exploring simplifying action-related types a little bit and I think that this should fix this but I have some other stuff to do first so I'm not sure when I will get back to it.

**Reply by tkofh** [reply]
`strictFunctionTypes` appears to be the culprit in the example I sent, disabling that makes the problem go away. 

One slight variation, possibly not related, [this codesandbox](https://codesandbox.io/s/xstate-createmodel-invoke-assign-empty-event-hr3eb?file=/src/index.ts) has an empty event assigner and that brings the error back even without strict mode enabled.

For now, asserting any non-string transition as `any` makes the error go away. Sometimes `// @ts-ignore`-ing the `invoke` key does the trick, but sometimes typescript thinks the entire machine config is bad.

**Reply by Andarist** [reply]
> strictFunctionTypes appears to be the culprit in the example I sent, disabling that makes the problem go away.

Well, I wouldn't call it a culprit per se :P it errors correctly for given types 😉 

> One slight variation, possibly not related, this codesandbox has an empty event assigner and that brings the error back even without strict mode enabled.

This one is caused by this incorrect declaration:
https://github.com/davidkpiano/xstate/blob/2f428bd70395d1ddaaaa8061d66f6f66492a215f/packages/core/src/types.ts#L445-L447
It expects `data` property on your `error.platform.*` event.

**Reply by tkofh** [reply]
Yeah culprit is probably a strong word here 😄 

When you say "this incorrect declaration", are you saying this is considered an issue inside xstate? I can totally see an argument for requiring a "reason" for the error in the error event, maybe my example just doesn't make sense.

**Reply by Andarist** [reply]
> When you say "this incorrect declaration", are you saying this is considered an issue inside xstate?

Yes - `onError` shouldn't be typed using `DoneInvokeEvent` as `onError` transitions are currently triggered by `error.platform.*` events and not `done.invoke.*` ones.
